### DISCUSSION TITLE
Stricter/typesafe `parse*` functions?

### DISCUSSION THREAD

**Discussion by julienfouilhe** [colinhacks/zod#1335] [Ideas]
I was wondering if it could be beneficial to add strict and typesafe parse functions directly into this library?

My usecase is this:

I use transforms and refinements to parse a gRPC input into domain objects, here is an example:

```
const schema = z.object({
  uid: z.string().transform(User.Id.into),
  // ... I also have stricter parse methods add context issues when an business error happens
});
```

Which makes the input and the output types different: Input is ` { uid: string }` while Output is `{ uid: User.Id.T }`

Then I use this on my gRPC input

```
parsed = schema.safeParseAsync(call.request)
```

But `call.request`, since I'm using gRPC, is already typed as such:

```
{
  uid: string;
}
```

It would be great if there could be a stricter method than `safeParseAsync` which would not accept unknown as an argument, but rather the Input of the schema, so that if my schema input is not compatible with the gPRC input, it gives me a type error instead of a runtime error.

For instance I would like a type error if I make a typo in my schema and I use `id` instead of `uid`

I would be happy to contribute these changes but I just want to make sure it makes sense to this library's contributors.

**Comment by scotttrinh** [comment]
If I understand this correctly, you want something that basically can trust the input but runs the transforms and refinements, right? Seems useful and adds a nice performance benefit in cases where you want the transformation pipeline for consistency, but don't need the input validation.

**Reply by scotttrinh** [reply]
Just dropping a note here that we should also consider the `default` "effect", too.

**Reply by mwisnicki** [reply]
I actually see the need to consider default but ignore transforms.
Imagine I want to create an object in the code rather than parsing - I can already provide values of output types so no need for transforms however I would like to take advantage of defaults.

```ts
const Schema = z.object({
  foo: z.string().transform(parseInt),
  bar: z.number().default(123)
});
const SchemaConstructor = z.constructor<typeof Schema>; // { foo: number, bar?: number } 
Schema.construct({ foo: 1 }) // {foo: 1, bar: 123}
```

**Reply by julienfouilhe** [reply]
@scotttrinh [I created a PR](https://github.com/colinhacks/zod/pull/1342) to do what you proposed, I think. I had only thought about the build time without thinking it meant zod could maybe bypass the runtime checks.
Would be happy to get feedback, if you're keen.
